### PR TITLE
Use nvjpeg2k ROI API directly

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

## Category:

**Refactoring**

## Description:

- It reworks nvjpeg2k decoder to use the higher level ROI decoding API directly, instead of decoding tile by tile

### Affected modules and functionalities:

Experimental image decoder

### Tests:

- [x] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [x] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-3249

### CODE REVIEW

File: dali/imgcodec/decoders/nvjpeg2k/nvjpeg2k.cc [hunk 1]
```diff
@@ -237,31 +173,47 @@ DecodeResult NvJpeg2000DecoderInstance::DecodeImplTask(int thread_idx,
...
-    (ctx.bpp != 8 && ctx.bpp != 16);
+    (ctx.bpp != 8 && ctx.bpp != 16) ||
+    is_roi_oob;
 
   auto decode_out = out;
+  auto decode_sh = ctx.orig_shape;
+  ROI roi_decode;  // ROI passed to decode API
+  ROI roi_convert;  // ROI passed to convert API
+  if (ctx.roi && !is_roi_oob) {
+    // if ROI decoding and ROI withing bounds
```
Comment (stiepan) [inline dali/imgcodec/decoders/nvjpeg2k/nvjpeg2k.cc:194]:
```suggestion
    // if ROI decoding and ROI within bounds
```

File: dali/imgcodec/decoders/nvjpeg2k/nvjpeg2k.cc [hunk 2]
```diff
@@ -237,31 +173,47 @@ DecodeResult NvJpeg2000DecoderInstance::DecodeImplTask(int thread_idx,
...
   if (is_processing_needed) {
     int64_t type_size = dali::TypeTable::GetTypeInfo(ctx.pixel_type).size();
-    size_t new_size = volume(ctx.shape) * type_size;
-    if (new_size > res.intermediate_buffer.capacity()) {
-      CUDA_CALL(cudaStreamSynchronize(ctx.cuda_stream));
-    }
-    res.intermediate_buffer.resize(volume(ctx.shape) * type_size);
-    decode_out = {res.intermediate_buffer.data(), ctx.shape, ctx.pixel_type};
+    size_t nbytes = volume(decode_sh) * type_size;
+    void *tmp_buff = scratchpad.AllocateGPU<uint8_t>(nbytes);
```
Comment (stiepan) [inline dali/imgcodec/decoders/nvjpeg2k/nvjpeg2k.cc:208]: Are we fine with alignment issues for `uint16_t` here (`.AllocateGPU<uint8_t>`)?

File: dali/imgcodec/decoders/nvjpeg2k/nvjpeg2k.cc [hunk 3]
```diff
@@ -237,31 +173,47 @@ DecodeResult NvJpeg2000DecoderInstance::DecodeImplTask(int thread_idx,
...
   bool is_processing_needed =
     channels > 1 ||  // nvJPEG2000 decodes into planar layout
     ctx.pixel_type != opts.dtype ||
     format != opts.format ||
-    (ctx.bpp != 8 && ctx.bpp != 16);
+    (ctx.bpp != 8 && ctx.bpp != 16) ||
+    is_roi_oob;
 
   auto decode_out = out;
+  auto decode_sh = ctx.orig_shape;
```
Comment (stiepan) [inline dali/imgcodec/decoders/nvjpeg2k/nvjpeg2k.cc:190]: nitpick, the assigned value is never used - it gets overwritten in both branches of the following if.

**Review by banasraf — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

**Review by stiepan — APPROVED** [review]

**Pull request merged**